retry signing after current signing session fails

BatiGencho opened the pull request.

arminsabouri commented on bin/test-suite/src/suite/consensus/frost/test_frost_e2e.rs:
```diff
@@ -324,3 +326,248 @@ pub async fn frost_e2e(suite: &ConsensusIntegrationTestSuite) -> Result<(), supe
+    tokio::time::sleep(Duration::from_secs(10)).await;
+
+    // now reconnect the peers of fed member 0
+    test_fed_members.get(&0).cloned().unwrap().send_test_signal(TestSignal::DisconnectAll());
```
The comment says reconnect but the signal is disconnect

BatiGencho replied:
correct, sorry. I have fixed it now.

arminsabouri commented on bin/test-suite/src/suite/consensus/frost/test_frost_e2e.rs:
```diff
@@ -324,3 +326,248 @@ pub async fn frost_e2e(suite: &ConsensusIntegrationTestSuite) -> Result<(), supe
+        mint_contract.burn(pegout_destination, pegout_data, pegout_amount.to_wei()).await.unwrap();
+    it_info_print!("Pegout Tx Receipt: ", tx_receipt);
+
+    // ===================== FAILURE =====================
```
Should this reconnect / disconnect be happening after L521?
Where we send the final eoa tx?

BatiGencho replied:
I have re-arranged this logic now.

arminsabouri commented on crates/consensus/authority/src/signing.rs:
```diff
@@ -500,13 +610,52 @@ where
+        // the coordinator is always expected to be us in this case, i.e. None
+        let coordinator = self.get_coordinator().await?;
+        if coordinator.is_some() {
+            panic!(">>>>>>>>>>> A non-coordinator is trying to (re)initiate a signing process!");
```
Should we be panicing here? What If this is just a timing error?

BatiGencho replied:
i made an error out of it. It should not happen though

arminsabouri commented on crates/consensus/authority/src/utils.rs:
```diff
@@ -495,12 +520,16 @@ pub(crate) enum GenerateSigningSesssionIdError {
     authority_pk: &secp256k1::PublicKey,
+    has_timestamp: bool,
 ) -> Result<SigningSessionId, GenerateSigningSesssionIdError> {
     let mut engine = sha256::HashEngine::default();
```
I feel like this function ended up getting more complicated than it needs to for its intended usecase.
How do you feel about just hashing uuidv4()

BatiGencho replied:
done. Much cleaner

arminsabouri commented on crates/consensus/common/src/utils.rs:
```diff
@@ -295,6 +295,23 @@ pub fn is_inturn(authorities_len: u64, signer_index: u64) -> bool {
 }
 
+/// Returns the inturn interval for a signer index
+pub fn get_in_turn_interval(authorities_len: u64, signer_index: u64) -> (u64, u64, u64, u64) {
```
Can we typedef `(u64, u64, u64, u64)` to somethign like `CoordinatorInterval`?

BatiGencho replied:
done. Thx for the suggestion

arminsabouri committed:
added mechanisms to compensate failed signing

* refactor(btc_server): remove unused remove utxo endpoint

* fix(btc_server): remove unused `remove_utxo()` db function

* fixed pipelines and make fmt

* minor fix

---------

Co-authored-by: Evgeni Pirianov <evgeni.pirianov@gmail.com>

fixed pipelines

retry signing after current signing session fails

fixed int. pipelines

fixed int. pipelines

scottmillner commented on crates/consensus/authority/src/signing.rs:
```diff
@@ -545,49 +695,99 @@ where
+                return Ok(());
+            }
+
+            // session exists and is valid, return if we are not in round 2
```
nit: should say if we are not in round 1

it confused me for a second lol

scottmillner reviewed (dismissed):
as always 🔥 🔥 🔥 🔥 🔥 
awesome sauce!!

arminsabouri commented on crates/net/network/src/frost/protocol.rs:
```diff
@@ -163,6 +163,7 @@ impl Stream for FrostProtoConnection {
             return Poll::Ready(Some(initial_ping.encoded()))
         }
         let peer_message_forwarder = this.peer_message_forwarder.clone();
+        //loop {
```
Nit: remove commented out code

BatiGencho replied:
sure, done

arminsabouri commented on crates/consensus/authority/src/signing.rs:
```diff
@@ -144,43 +165,93 @@ where
+    pub(crate) fn insert_new_signing_session(
         &mut self,
         session_id: [u8; 32],
+        coordinator_id: u64,
```
nit: is this an index into the fed list? if so it should be `coordiantor_idnex`

BatiGencho replied:
of course, fixed

arminsabouri commented on crates/consensus/authority/src/signing.rs:
```diff
@@ -504,13 +614,53 @@ where
+        let (start, end, time_passed, time_remaining) =
+            self.get_inturn_interval_for_coordinator(None);
+
+        // try to find the signing session in cache in case it was re-triggered by the same
+        // coordinator
+        match self.get_signing_session(session_id) {
```
What actually causes this to be re-triggered. IIRC the block builder just waits 60s for the current signing session to pass?

BatiGencho replied:
check the logic here: https://github.com/botanix-labs/Macbeth/pull/168/files#diff-f459d49b88e4803ede826537533f5067794f1d39165d9791add7a6bb6b3ac777R1017-R1067

if the current coordinator is still in turn and it decides there is time, it would retry it.

arminsabouri reviewed (commented):
Awesome work. Just had a few questions

arminsabouri commented on .github/workflows/unit.yml:
```diff
@@ -20,52 +20,6 @@ concurrency:
-          cache-on-failure: true
-      - name: Run e2e tests
-        run: |
-          make start-test-suite
```
Did we mean to delete this? Is this being run somewhere else?

arminsabouri replied:
I just saw this was moved to unit.yml
Closing

arminsabouri commented on bin/test-suite/src/suite/consensus/frost/test_frost_e2e_edge_cases.rs:
```diff
@@ -0,0 +1,294 @@
+    it_info_print!("Eoa Tx Receipt: ", eoa_tx_receipt);
+
+    // ===================== FAILURE =====================
+    // wait for a few seconds to allow the signing to start ???
+    tokio::time::sleep(Duration::from_secs(3)).await;
+
+    // now disconnect the peers of fed member 0
+    test_fed_members.get(&0).cloned().unwrap().send_test_signal(TestSignal::DisconnectAll());
+
+    // wait for a new epoch to start
+    tokio::time::sleep(Duration::from_secs(10)).await;
+
+    // now reconnect the peers of fed member 0
+    test_fed_members.get(&0).cloned().unwrap().send_test_signal(TestSignal::ReconnectAll());
+    tokio::time::sleep(Duration::from_secs(3)).await;
```
What are we testing here? That we can perform a signing round without a peer being connected?

arminsabouri replied:
It would be helpful to have a comment above the test function describing what exact edge case we are testing

BatiGencho replied:
Sure, I will add more comments in the next commit!

arminsabouri commented:
Don't mean to be a commit Nazi. But the commits here could probably be squashed into one or maybe two 
- signing edge cases
- Fixed int pipeline

arminsabouri commented on crates/botanix-lib/src/extra_data_header.rs:
```diff
@@ -368,11 +368,28 @@ mod tests {
-        let serialized =
-            ExtraDataHeader::deserialize(&mut buf.as_slice()).expect("Deserialization");
-        assert_eq!(serialized, header);
+        // Check version
```
I simplified this test in a earlier commit b/c I don't think doing exact bit by bit matching really got us much.
I think testing if something deserializes and serializes to the same thing is suffecient.
Lmk if you disagree

BatiGencho replied:
I totally agree, all good!
